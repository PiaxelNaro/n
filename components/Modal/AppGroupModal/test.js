// @flow
import React from 'react';
import { shallow } from 'enzyme';
import { _AppGroupModal as AppGroupModal } from '.';
import mockI18n from '../../../modules/i18n/mockI18n';

describe('<AppGroupModal>', () => {
    it('handles group loading lifecycle', () => {
        let successCallback = jest.fn();
        let failureCallback = jest.fn();
        const fakePromise: any = {
            then: (success: Function, failure: Function) => {
                successCallback = success;
                failureCallback = failure;
            },
        };
        const modal = shallow(
            <AppGroupModal
                apps={[{ id: 0 }]}
                onRequestClose={jest.fn()}
                groupsPromise={fakePromise}
                createGroup={jest.fn()}
                addAppsToGroup={jest.fn()}
                i18n={mockI18n}
            />
        );
        modal.instance().componentDidMount();

        // 1. Component in loading state
        expect(modal.find('Spinner').exists()).toBeTruthy();

        // 2. Component in failure state
        failureCallback();
        expect(modal.find('AppGroupModal__ErrorWrapper').exists()).toBeTruthy();

        // 3. Component in success mode
        successCallback([
            {
                id: 1,
                name: 'Group 1',
                link: '/group-1',
                apps: [],
            },
        ]);
        expect(modal.find('AppGroup').exists()).toBeTruthy();
    });

    it('allows creating new groups', () => {
        let successCallback = jest.fn();
        const createGroupFn = jest.fn(() => ({
            then: success => {
                successCallback = success;
            },
        }));
        const fakePromise: any = {
            then: cb => {
                cb([
                    {
                        id: 1,
                        name: 'Group 1',
                        link: '',
                        apps: [],
                    },
                ]);
            },
        };
        const modal = shallow(
            <AppGroupModal
                apps={[{ id: 0 }]}
                onRequestClose={jest.fn()}
                groupsPromise={fakePromise}
                createGroup={createGroupFn}
                addAppsToGroup={jest.fn()}
                i18n={mockI18n}
            />
        );
        modal.instance().componentDidMount();
        expect(modal).toMatchSnapshot();

        modal.find('.js-create-group').simulate('click');
        modal.update();
        expect(modal).toMatchSnapshot();

        const input = modal.find('AppGroupModal__StyledInputWrapper');

        // No-op if the input is empty
        input.simulate('keyPress', { key: 'Test' });
        expect(createGroupFn).not.toHaveBeenCalled();

        input.simulate('keyPress', { key: 'Enter', target: { value: 'test' } });
        expect(createGroupFn).toHaveBeenCalledWith('test');

        // Loading state
        modal.update();
        let groups = modal.find('AppGroup');
        expect(groups.length).toEqual(2);
        expect(groups.at(0).prop('isLoading')).toBe(false);
        expect(groups.at(1).prop('isLoading')).toBe(true);

        // Success state
        successCallback({
            id: 2,
            name: 'Group 2',
            link: '',
            apps: [],
        });
        modal.update();
        groups = modal.find('AppGroup');
        expect(groups.length).toEqual(2);
        expect(groups.at(0).prop('isLoading')).toBe(false);
        expect(groups.at(1).prop('isLoading')).toBe(false);
    });

    it('handles failure to create group', () => {
        let errorCallback = jest.fn();
        const createGroupFn = jest.fn(() => ({
            then: (success, error) => {
                errorCallback = error;
            },
        }));
        const fakePromise: any = {
            then: cb => {
                cb([
                    {
                        id: 1,
                        name: 'Group 1',
                        link: '',
                        apps: [],
                    },
                ]);
            },
        };
        const modal = shallow(
            <AppGroupModal
                apps={[{ id: 0 }]}
                onRequestClose={jest.fn()}
                groupsPromise={fakePromise}
                createGroup={createGroupFn}
                addAppsToGroup={jest.fn()}
                i18n={mockI18n}
            />
        );
        modal.instance().componentDidMount();

        modal.find('.js-create-group').simulate('click');
        modal.update();
        const input = modal.find('AppGroupModal__StyledInputWrapper');
        input.simulate('keyPress', { key: 'Enter', target: { value: 'test' } });
        expect(createGroupFn).toHaveBeenCalledWith('test');

        // Error state
        errorCallback();
        modal.update();
        expect(modal.find('AppGroup').length).toEqual(1);
        expect(modal.find('AppGroupModal__ErrorWrapper').exists()).toBe(true);
    });

    it('allows adding apps to a group', () => {
        let successCallback = jest.fn();
        let errorCallback = jest.fn();
        const addAppsToGroup = jest.fn(() => ({
            then: (success, error) => {
                successCallback = success;
                errorCallback = error;
            },
        }));
        const fakePromise: any = {
            then: cb => {
                cb([
                    {
                        id: 1,
                        name: 'Group 1',
                        link: '',
                        apps: [],
                    },
                    {
                        id: 2,
                        name: 'Group 2',
                        link: '',
                        apps: [],
                    },
                ]);
            },
        };
        const modal = shallow(
            <AppGroupModal
                apps={[{ id: 0 }]}
                onRequestClose={jest.fn()}
                groupsPromise={fakePromise}
                createGroup={jest.fn()}
                addAppsToGroup={addAppsToGroup}
                i18n={mockI18n}
            />
        );
        modal.instance().componentDidMount();

        modal.find('AppGroup').at(0).prop('addToGroup')(1, [0]);
        expect(addAppsToGroup).toHaveBeenCalledWith(1, [0]);
        expect(modal.find('AppGroup').at(0).prop('group').apps.length).toEqual(0);

        // Loading state
        modal.update();
        expect(modal.find('AppGroup').at(0).prop('isLoading')).toBe(true);
        expect(modal.find('AppGroup').at(1).prop('isLoading')).toBe(false);

        // Error state
        errorCallback();
        modal.update();
        expect(modal.find('AppGroup').at(0).prop('isLoading')).toBe(false);
        expect(modal.find('AppGroupModal__ErrorWrapper').exists()).toBe(true);

        // Success state
        successCallback({
            '1': {
                id: 1,
                name: 'Group 1',
                link: '',
                apps: [{ id: 0 }],
            },
        });
        modal.update();
        expect(modal.find('AppGroup').at(0).prop('isLoading')).toBe(false);
        expect(modal.find('AppGroup').at(0).prop('group').apps.length).toEqual(1);
        expect(modal.find('AppGroup').at(1).prop('group').apps.length).toEqual(0);
    });
});
