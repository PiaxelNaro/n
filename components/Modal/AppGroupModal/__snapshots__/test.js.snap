// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppGroupModal> allows creating new groups 1`] = `
<Modal
  contentLabel=""
  isOpen={true}
  onAfterOpen={[Function]}
  onRequestClose={[Function]}
>
  <ModalHeader
    onRequestClose={[Function]}
    title="Add to your group"
  />
  <BasicModal__ModalBody>
    <AppGroupModal__StyledModalBody>
      <AppList
        apps={
          Array [
            Object {
              "id": 0,
            },
          ]
        }
        i18n={
          Object {
            "gettext": [Function],
            "lang": "en-US",
            "ngettext": [Function],
            "npgettext": [Function],
            "pgettext": [Function],
          }
        }
      />
      <FlexGrid.Item
        flex={1}
      >
        <AppGroupModal__CenteredModalBody
          gutter="5px"
        >
          <Spinner
            size={16}
          />
          <span>
            Loading data
          </span>
        </AppGroupModal__CenteredModalBody>
      </FlexGrid.Item>
    </AppGroupModal__StyledModalBody>
  </BasicModal__ModalBody>
  <BasicModal__ModalFooter>
    <FlexGrid>
      <FlexGrid.Item
        flex={1}
      >
        <AppGroupModal__CreateGroupButton
          className="js-create-group"
          onClick={[Function]}
        >
          <FlexGrid
            gutter="3px"
          >
            <MdControlPoint
              size={16}
            />
            <span>
              Create a new group
            </span>
          </FlexGrid>
        </AppGroupModal__CreateGroupButton>
      </FlexGrid.Item>
      <ButtonPrimary
        onClick={[Function]}
      >
        Done
      </ButtonPrimary>
    </FlexGrid>
  </BasicModal__ModalFooter>
</Modal>
`;

exports[`<AppGroupModal> allows creating new groups 2`] = `
<Modal
  contentLabel=""
  isOpen={true}
  onAfterOpen={[Function]}
  onRequestClose={[Function]}
>
  <ModalHeader
    onRequestClose={[Function]}
    title="Add to your group"
  />
  <BasicModal__ModalBody>
    <AppGroupModal__StyledModalBody>
      <AppList
        apps={
          Array [
            Object {
              "id": 0,
            },
          ]
        }
        i18n={
          Object {
            "gettext": [Function],
            "lang": "en-US",
            "ngettext": [Function],
            "npgettext": [Function],
            "pgettext": [Function],
          }
        }
      />
      <AppGroupModal__StyledGroupListWrapper>
        <AppGroup
          addToGroup={[Function]}
          appIds={
            Array [
              0,
            ]
          }
          group={
            Object {
              "apps": Array [],
              "id": 1,
              "link": "",
              "name": "Group 1",
            }
          }
          i18n={
            Object {
              "gettext": [Function],
              "lang": "en-US",
              "ngettext": [Function],
              "npgettext": [Function],
              "pgettext": [Function],
            }
          }
          isLoading={false}
        />
      </AppGroupModal__StyledGroupListWrapper>
    </AppGroupModal__StyledModalBody>
  </BasicModal__ModalBody>
  <BasicModal__ModalFooter>
    <ClickOutside
      onClickOutside={[Function]}
    >
      <AppGroupModal__StyledInputWrapper
        autoFocus={true}
        maxLength="50"
        onKeyPress={[Function]}
        placeholder="Give your group a name and press enter"
      />
    </ClickOutside>
  </BasicModal__ModalFooter>
</Modal>
`;
